src/coprocessor/time_limiter.rs

```diff
@@ -0,0 +1,171 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use pin_project::pin_project;
+use std::future::Future;
+use std::marker::PhantomData;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::{Duration, Instant};
+use tokio::sync::{Semaphore, SemaphorePermit};
+
+use super::metrics::*;
+
+/// Limits the time used by `fut` before a semaphore permit is acquired.
+///
+/// The future `fut` can always run for at least `time_limit_without_permit`,
+/// but it needs to acquire a permit from the semaphore before it can continue.
+pub fn limit_time<'a, F: Future + 'a>(
+    fut: F,
+    semaphore: &'a Semaphore,
+    time_limit_without_permit: Duration,
+) -> impl Future<Output = F::Output> + 'a {
+    TimeLimiter::new(semaphore.acquire(), fut, time_limit_without_permit)
+}
+
+#[pin_project]
+struct TimeLimiter<'a, PF, F>
+where
+    PF: Future<Output = SemaphorePermit<'a>>,
+    F: Future,
+{
+    #[pin]
+    permit_fut: PF,
+    #[pin]
+    fut: F,
+    time_limit_without_permit: Duration,
+    execution_time: Duration,
+    permit: Option<SemaphorePermit<'a>>,
+    // Whether the task is waiting for a semaphore. Used for tracking metrics only.
+    waiting: bool,
```

Comment:
We can make them into a state enum: Limitation: NotLimited, Acquiring, Acuqired(SemaphorePermit)